### PR TITLE
OSD-22892 (Part 4): Probe Integration

### PR DESCRIPTION
**Pull request by abyrne55**
This PR is composed of commits cherry-picked from my larger [OSD-22892](https://issues.redhat.com/browse/OSD-22892) branch for the sake of readability. This PR covers integration of the new Probe interface into the verifier's critical path for AWS. IOW, this PR enables all of the code that was added in Parts 1-3 (#240, #241, and #245). **Unlike previous PRs in this series, please review the code AND test it out on your local machine when reviewing this PR.**

### Highlights
* All egress tests are now routed through an implementation of the [new Probe interface](https://github.com/openshift/osd-network-verifier/blob/main/pkg/probes/package_probes.go) instead of the old tightly-coupled functions built into the `awsverifier` package
  * The "old ways" (i.e., launching an instance of the golden AMI which spins up a docker image...) can still be used via the LegacyProbe (see #241)
* [_breaking API change_] The `ValidateEgressInput` struct now contains a new mandatory `Probe` field (type: `probes.Probe`) that allows the caller to dictate which Probe implementation (i.e., `curl_json.CurlJSONProbe`, `legacy.LegacyProbe`, or an external/user-provided Probe implementation) to use when probing the target network
  * The undocumented and temporary `FeatureFlags` field has been removed; this field was only ever used to implement the experimental version of the probe interface
* The `osd-network-verifier` CLI tool now takes an optional `--probe` flag allowing CLI users to select a Probe implementation by name (e.g., `--probe=legacy`); defaults to CurlJSONProbe
  * The `integration` CLI testing tool also now handles `--probe`, but as a _required_ flag
* [_external API change_] `awsverifier.GetAMIForRegion()` has been deprecated (see table below)
  * Current callers of this function (just osdctl and CAD, AFAICT) can just allow the verifier to auto-select an AMI by not passing a value for `CloudImageId` when calling `ValidateEgress()`

This PR removes/relocates a lot of old code/resources that have been obsoleted by previous PRs in this series, as summarized by the table below.
| Item | Old Location | New Location | How to Access |
| ----------- | ----------- | ----------- | ----------- |
| userdata templates | `helpers` package | their parent Probe's package | `Probe.GetExpandedUserdata()` |
| AMI mappings | aws_verifier.go | their parent Probe's package (machine_images.go) | `Probe.GetMachineImageID()` or just allow auto-selection (see above) |
| egress URL lists | golden AMI or experimental_probe.go | embedded* in `data/egress_lists` package | `egress_lists.GetEgressListAsString()` |
| console output parsing logic | `AwsVerifier.findUnreachableEndpoints()` | Probe implementations | `Probe.ParseProbeOutput()` |

*transitional implementation; the API for accessing egress lists will be finalized alongside the list downloading feature ([OSD-22628](https://issues.redhat.com/browse/OSD-22628))

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation* and example code
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have tested the functionality against ~~gcp /~~ aws, it doesn't cause any regression
- [x] I have added execution results to the PR's readme

*This PR just updates the integration test's docs. Updates to `/docs` and the main README will come in a future PR

## How to test this PR
Checkout this PR, `cd` into its root, and run `make test build`. Play around with `./osd-network-verifier` as you normally would, adding the `--probe=curl` or `--probe=legacy` flag, e.g.
```sh
./osd-network-verifier egress --debug --profile=default --region=us-east-1 --subnet-id=subnet-01335384c9052ac3e --probe=curl --timeout 4s
```

Also, run the integration test: `cd` into `integration`, run `go mod tidy` and `go build`, then run something along the lines of:
```sh
./integration --region=us-east-2 --profile=$YOUR_AWS_PROFILE_NAME --probe=curl # or --probe=legacy
```

Bonus points: spin up [a transparently-proxied VPC](https://github.com/abyrne55/transparent-proxy-setup/) and test this PR against that
```
git clone https://github.com/abyrne55/transparent-proxy-setup.git
cd transparent-proxy-setup
# Follow https://github.com/abyrne55/transparent-proxy-setup/#setup
# Once you're able to curl the CA cert, run the following
/path/to/this/PR/osd-network-verifier egress --profile=default --subnet-id=$(terraform output -raw proxied_subnet_id) --region=$(terraform output -raw region) --cacert=cacert.pem --probe=curl
```

## Logs 

<details><summary>Click to show sample integration test output</summary>

```
$ ./integration --region=us-east-1 --profile=default --probe=legacy
2024/07/05 17:52:48 using region us-east-1
2024/07/05 17:52:48 using AZ: us-east-1a
2024/07/05 17:52:49 waiting up to 10s for vpc to become available
2024/07/05 17:52:49 created VPC: vpc-0161475326f557bac
2024/07/05 17:52:49 enableDnsSupport is true for VPC: vpc-0161475326f557bac
2024/07/05 17:52:50 enableDnsHostnames is true for VPC: vpc-0161475326f557bac
2024/07/05 17:52:50 creating private subnet: 10.0.0.0/25
2024/07/05 17:52:50 waiting up to 1m0s for subnet to become available
2024/07/05 17:52:50 subnet available: subnet-0ba3e8f671888093d
2024/07/05 17:52:50 created private subnet: subnet-0ba3e8f671888093d
2024/07/05 17:52:50 creating firewall subnet: 10.0.0.128/25
2024/07/05 17:52:51 waiting up to 1m0s for subnet to become available
2024/07/05 17:52:51 subnet available: subnet-0c162077042bed176
2024/07/05 17:52:51 created firewall subnet: subnet-0c162077042bed176
2024/07/05 17:52:51 creating public subnet: 10.0.1.0/25
2024/07/05 17:52:51 waiting up to 1m0s for subnet to become available
2024/07/05 17:52:52 subnet available: subnet-02a8fe73eaee167dc
2024/07/05 17:52:52 created public subnet: subnet-02a8fe73eaee167dc
2024/07/05 17:52:52 created firewall rule group: arn:aws:network-firewall:us-east-1:557801144153:stateful-rulegroup/osd-network-verifier-rule-group
2024/07/05 17:52:53 created firewall policy: arn:aws:network-firewall:us-east-1:557801144153:firewall-policy/osd-network-verifier-firewall-policy
2024/07/05 17:52:53 created firewall: arn:aws:network-firewall:us-east-1:557801144153:firewall/osd-network-verifier-firewall
2024/07/05 17:52:54 allocated EIP address: eipalloc-0bd8de3d312988cae
2024/07/05 17:52:54 created NAT gateway: nat-06d88cd48caeb08f3
2024/07/05 17:52:54 created internet gateway: igw-0c003548de51a00a9
2024/07/05 17:52:55 created private subnet route table: rtb-07253ce00bba674b3
2024/07/05 17:52:55 associated private subnet route table: rtbassoc-04b4bf8819b6827ad
2024/07/05 17:52:55 created public subnet route table: rtb-01fddcb18f2589de3
2024/07/05 17:52:55 associated public subnet route table: rtbassoc-03dfbadf103ece8a8
2024/07/05 17:52:56 created firewall subnet route table: rtb-0b33e5606b9ef3c5c
2024/07/05 17:52:56 associated firewall subnet route table: rtbassoc-0cddc17b5b09c55d5
2024/07/05 17:52:56 waiting up to 30s for internet gateway attachment to become available
2024/07/05 17:52:56 internet gateway attached: igw-0c003548de51a00a9
2024/07/05 17:52:56 route created for firewall subnet 0.0.0.0/0 --> IGW
2024/07/05 17:52:57 created internet gateway route table: rtb-0b33e5606b9ef3c5c
2024/07/05 17:52:57 associated internet gateway route table: rtbassoc-023fa3bcb25358324
2024/07/05 17:52:57 waiting up to 5m0s for the NAT Gateway to become available
2024/07/05 17:55:01 NAT gateway available: nat-06d88cd48caeb08f3
2024/07/05 17:55:02 route created for private subnet 0.0.0.0/0 --> NAT
2024/07/05 17:55:02 waiting up to 10m0s for the firewall to become READY
2024/07/05 17:55:02 status: PROVISIONING
2024/07/05 17:55:17 status: PROVISIONING
2024/07/05 17:55:37 status: PROVISIONING
2024/07/05 17:56:05 status: PROVISIONING
2024/07/05 17:58:03 status: READY
2024/07/05 17:58:03 firewall ready: arn:aws:network-firewall:us-east-1:557801144153:firewall/osd-network-verifier-firewall
2024/07/05 17:58:03 route created for public subnet 0.0.0.0/0 --> firewall
2024/07/05 17:58:04 Starting ONV egress validation
Created security group with ID: sg-03a72a7f7746f0438
Created instance with ID: i-0f573b5d1eaea72b0
Deleting instance with ID: i-0f573b5d1eaea72b0
Deleting security group with ID: sg-03a72a7f7746f0438
Summary:
printing out failures:
 - egressURL error: cdn03.quay.io:443
 - egressURL error: quay.io:443
 - egressURL error: cdn01.quay.io:443
 - egressURL error: cdn02.quay.io:443

printing out exceptions preventing the verifier from running the specific test:
printing out errors faced during the execution:
2024/07/05 18:01:07 egress failure: cdn03.quay.io:443
2024/07/05 18:01:07 egress failure: quay.io:443
2024/07/05 18:01:07 egress failure: cdn01.quay.io:443
2024/07/05 18:01:07 egress failure: cdn02.quay.io:443
2024/07/05 18:01:07 ONV egress validation: Failure!
2024/07/05 18:01:07 cleaning up route table: rtb-01a4d71ffc6b80e10
2024/07/05 18:01:08 disassociated association: rtbassoc-023fa3bcb25358324
2024/07/05 18:01:08 deleted route table: rtb-01a4d71ffc6b80e10
2024/07/05 18:01:08 cleaning up route table: rtb-0b33e5606b9ef3c5c
2024/07/05 18:01:08 disassociated association: rtbassoc-0cddc17b5b09c55d5
2024/07/05 18:01:08 deleted route table: rtb-0b33e5606b9ef3c5c
2024/07/05 18:01:08 cleaning up route table: rtb-07253ce00bba674b3
2024/07/05 18:01:09 disassociated association: rtbassoc-04b4bf8819b6827ad
2024/07/05 18:01:09 deleted route table: rtb-07253ce00bba674b3
2024/07/05 18:01:09 cleaning up route table: rtb-01fddcb18f2589de3
2024/07/05 18:01:09 disassociated association: rtbassoc-03dfbadf103ece8a8
2024/07/05 18:01:09 deleted route table: rtb-01fddcb18f2589de3
2024/07/05 18:01:10 deleting firewall osd-network-verifier-firewall
2024/07/05 18:01:10 waiting up to 20m0s for the firewall to be deleted
2024/07/05 18:01:10 status: DELETING
2024/07/05 18:01:25 status: DELETING
2024/07/05 18:01:44 status: DELETING
2024/07/05 18:02:16 status: DELETING
2024/07/05 18:03:57 firewall deleted: osd-network-verifier-firewall
2024/07/05 18:03:58 deleting firewall policy osd-network-verifier-firewall-policy
2024/07/05 18:03:58 waiting up to 10m0s for the firewall policy to be deleted
2024/07/05 18:03:58 status: DELETING
2024/07/05 18:04:13 firewall policy deleted: osd-network-verifier-firewall-policy
2024/07/05 18:04:13 deleting firewall rule group osd-network-verifier-rule-group
2024/07/05 18:04:13 rule group deleted: osd-network-verifier-rule-group
2024/07/05 18:04:14 deleting NAT Gateway: nat-06d88cd48caeb08f3
2024/07/05 18:04:14 waiting up to 5m0s for the NAT Gateway to be deleted
2024/07/05 18:05:16 NAT gateway deleted: nat-06d88cd48caeb08f3
2024/07/05 18:05:17 EIP released: eipalloc-0bd8de3d312988cae
2024/07/05 18:05:17 detached internet gateway: igw-0c003548de51a00a9 from vpc: vpc-0161475326f557bac
2024/07/05 18:05:17 internet gateway deleted: igw-0c003548de51a00a9
2024/07/05 18:05:18 deleted subnet: subnet-0c162077042bed176
2024/07/05 18:05:19 deleted subnet: subnet-02a8fe73eaee167dc
2024/07/05 18:05:19 deleted subnet: subnet-0ba3e8f671888093d
2024/07/05 18:05:19 No SecurityGroups found - skipping cleanup
2024/07/05 18:05:20 deleted VPC: vpc-0161475326f557bac
```

</details>

### CODE REVIEW

**Review by fahlmant — APPROVED** [review]
/lgtm
/hold
Please feel free to unhold whenever, and consider adding the squash label as well

File: pkg/data/egress_lists/egress_lists.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+// URLs with tlsDisabled=false, while the second string contains all URLs with tlsDisabled=true
+func GetEgressListAsString(platformType string, region string) (string, string, error) {
+	normalizedPlatformType, err := helpers.GetPlatformType(platformType)
+	if err != nil {
+		return "", "", err
+	}
+
+	var egressListYamlStr string
+	switch normalizedPlatformType {
+	case helpers.PlatformGCP:
```
Comment (AlexVulaj) [inline pkg/data/egress_lists/egress_lists.go:39]: Without digging through this code too much - should we also check for the "new" platformtype name equivalents? i.e.:
```suggestion
	case helpers.PlatformGCP, helpers.PlatformGCPClassic:
```

**Comment by AlexVulaj** [comment]
/tide/merge-method-squash

File: pkg/data/egress_lists/egress_lists.go (hunk 1 above)
Comment (abyrne55) [reply pkg/data/egress_lists/egress_lists.go:39]: Thanks for the review! The `normalizedPlatformType, err := helpers.GetPlatformType(platformType)` a few lines up ensures that normalizedPlatformType will always be PlatformGCP, PlatformAWS, or PlatformHostedCluster. That being said, I'd love to talk about ripping out the old platform type values in a future PR so we can avoid this confusion entirely.

**Pull request merged**